### PR TITLE
upgrade: Add a validation check for postponed nodes.

### PR DESCRIPTION
**Pull request by jsuchome**
- When the upgrade is not finished, show a link to wizard.
- Add a validation check to make sure postponed nodes can be skipped.
- Let skip_unready_nodes skip also nodes that are in crowbar_upgrade state 

https://jira.prv.suse.net/browse/SCRD-3745

### CODE REVIEW

**Comment by susecloudcommitpolice** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please remove period from end of commit subject line.<br />
83d5aa7d2ed45039c068eecec2fa9508e5c4c2e5</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please remove period from end of commit subject line.<br />
7d2af29d3c05db741e7265cbd69509a35b2bf267</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: crowbar_framework/app/controllers/application_controller.rb [hunk 1]
```diff
@@ -34,6 +34,10 @@ class ApplicationController < ActionController::Base
     Rails.env.test?
   }
   before_filter :upgrade, if: proc {
+    if File.exist?("/var/lib/crowbar/upgrade/7-to-8-upgrade-compute-nodes-postponed")
+      flash[:warning] = "Some nodes are not upgraded yet. " \
```
Comment (jhesketh) [inline crowbar_framework/app/controllers/application_controller.rb]: are flash's translatable somewhere?

Comment (jsuchome) [reply crowbar_framework/app/controllers/application_controller.rb]: I'm not sure about flash messages, but validation ones are sometimes set for translation...
I chose not to this time, as other  validation strings in the same file are not translated.

It's a bit messy.

**Comment by susecloudcommitpolice** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please shorten commit subjects to less than 70 chars</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: crowbar_framework/app/models/service_object.rb [hunk 2]
```diff
@@ -1902,8 +1931,8 @@ def skip_unready_nodes(bc, inst, new_elements, old_elements)
         pre_cached_nodes[n] ||= Node.find_by_name(n)
         node = pre_cached_nodes[n]
         next if node.nil?
-        # skip if nodes are on ready or crowbar_upgrade state, we dont need to do anything
-        next if ["ready", "crowbar_upgrade"].include?(node.state)
+        # skip if nodes are on ready, we dont need to do anything
+        next if node.state == "ready"
```
[the remark this answers is not in the document]
Comment (rhafer) [reply crowbar_framework/app/models/service_object.rb]: @jsuchome Hm. Would it help to only skip the check for "crowbar_upgrade" if "7-to-8-upgrade-compute-nodes-postponed" exists?

[the remark this answers is not in the document]
Comment (jsuchome) [reply crowbar_framework/app/models/service_object.rb]: It would. But don't you find it super ugly?

[the remark this answers is not in the document]
Comment (rhafer) [reply crowbar_framework/app/models/service_object.rb]: Oh, yes for sure it's ugly! Unfortunately I am currently lacking any better idea. Might need to spend some additional thoughts on that.

You could dress it up a bit by introducing some library call for the check.

[the remark this answers is not in the document]
Comment (Itxaka) [reply crowbar_framework/app/models/service_object.rb]: cant we check for the actual upgrade step here?

`next if node.state == "crowbar_upgrade" && node["crowbar_upgrade_step"] == :services`

[the remark this answers is not in the document]
Comment (jsuchome) [reply crowbar_framework/app/models/service_object.rb]: We could, but that's just other way of going around it...

File: crowbar_framework/app/models/service_object.rb [hunk 3]
```diff
@@ -1884,6 +1918,11 @@ def skip_unchanged_nodes(elements, old_role, role)
     cleaned_elements
   end
 
+  # if this is true, node should not be skipped even though it's in crowbar_upgrade state
+  def crowbar_upgrade_needs_to_run(node)
```
Comment (rhafer) [inline crowbar_framework/app/models/service_object.rb]: Nit: I think this name is somewhat misleading. I'd suggest to reverse the logic here and call the method something like `upgrade_postponed?`

Also the `node` parameter is unused. Is it actually needed?

File: crowbar_framework/app/models/service_object.rb [hunk 4]
```diff
@@ -1902,8 +1941,9 @@ def skip_unready_nodes(bc, inst, new_elements, old_elements)
         pre_cached_nodes[n] ||= Node.find_by_name(n)
         node = pre_cached_nodes[n]
         next if node.nil?
-        # skip if nodes are on ready or crowbar_upgrade state, we dont need to do anything
-        next if ["ready", "crowbar_upgrade"].include?(node.state)
+        # skip if nodes are on ready, we dont need to do anything
+        next if node.state == "ready"
+        next if node.state == "crowbar_upgrade" && crowbar_upgrade_needs_to_run(node)
```
Comment (rhafer) [inline crowbar_framework/app/models/service_object.rb]: Some comment on why this is actually needed would be helpful IMO. Remember how long it took you to understand the logic.

File: crowbar_framework/app/models/service_object.rb (hunk 3 above)
Comment (jsuchome) [reply crowbar_framework/app/models/service_object.rb]: True about `node` and comments. 
`upgrade_postponed?` - OK, I can change it, but then I'll have to negate the output in the check later which is somewhat less nice.

File: crowbar_framework/app/models/service_object.rb [hunk 5]
```diff
@@ -682,9 +682,43 @@ def validate_proposal proposal
...
   end
 
+  #
+  # Make sure that when user wants to apply changes affecting the nodes that are postponed
+  # (= not upgraded while rest of the cloud is already upgraded)
+  # the nodes can be skipped.
+  #
+  def validate_postponed_nodes(proposal)
+    return unless upgrade_postponed?
+
```
Comment (Itxaka) [inline crowbar_framework/app/models/service_object.rb]: I would streamline and exit fast from here if there is nothing to do:

```
skip_unready_nodes = Rails.application.config.experimental.fetch("skip_unready_nodes", {})

skip_unready_nodes_enabled = skip_unready_nodes.fetch("enabled", false)
raise Crowbar::NeedOptionEnabled("skip_unready_nodes") unless skip_unready_nodes_enabled

skip_unready_nodes_roles = skip_unready_nodes.fetch("roles", [])
return if skip_unready_nodes_roles.size == 0

```

File: crowbar_framework/app/models/service_object.rb [hunk 6]
```diff
@@ -1647,6 +1681,11 @@ def run_remote_chef_client(node, command, logfile_name)
 
   THREAD_POOL_SIZE = 20
 
+  # Is the upgrade of (some) compute nodes postponed
+  def upgrade_postponed?
+    Dir.glob("/var/lib/crowbar/upgrade/*-upgrade-compute-nodes-postponed").any?
```
Comment (Itxaka) [inline crowbar_framework/app/models/service_object.rb]: should this point to the exact versioning for the upgrade (i.e. "7-8") or is that file not versioned yet?

File: crowbar_framework/app/models/service_object.rb [hunk 7]
```diff
@@ -682,9 +682,43 @@ def validate_proposal proposal
...
+      next if skip_unready_nodes_enabled && skip_unready_nodes_roles.include?(role)
+      nodes.each do |n|
+        next if is_cluster? n
+        node = NodeObject.find_by_name(n)
+        postponed_nodes << n unless node.upgraded?
+      end
+    end
+    return if postponed_nodes.empty?
+
+    validation_error "The upgrade of some nodes has been postponed: " +
```
Comment (skazi0) [inline crowbar_framework/app/models/service_object.rb]: won't it generate duplicates if there are multiple roles in one barclamp?

File: crowbar_framework/app/models/service_object.rb (hunk 5 above)
Comment (jsuchome) [reply crowbar_framework/app/models/service_object.rb]: > raise Crowbar::NeedOptionEnabled("skip_unready_nodes") unless skip_unready_nodes_enabled

Hm, there's no reason to raise here until we find out if any postponed node has a role in this barclamp.

> return if skip_unready_nodes_roles.size == 0

Why would we do it? Again, I think we must compare `skip_unready_nodes_roles` with the roles being applied.

File: crowbar_framework/app/models/service_object.rb (hunk 7 above)
Comment (skazi0) [reply crowbar_framework/app/models/service_object.rb]: I don't have any example at hand... but it doesn't mean that it can't happen :)

Comment (jsuchome) [reply crowbar_framework/app/models/service_object.rb]: I think it can't happen with our barclamps

Comment (jsuchome) [reply crowbar_framework/app/models/service_object.rb]: @skazi0 anyway, i've added `uniq` now.

File: crowbar_framework/app/controllers/application_controller.rb (hunk 1 above)
Comment (nicolasbock) [reply crowbar_framework/app/controllers/application_controller.rb]: Are flash message shown to the user in the UI? Then we might want to localize the message I would say.

Comment (jsuchome) [reply crowbar_framework/app/controllers/application_controller.rb]: That's the same as for validation strings - in the end, they are presented by flash messages.

And those defined in this file (the old ones, not the one I've added) are not set for translation...

**Pull request merged**